From: Michael Biebl <biebl@debian.org>
Date: Sat, 27 Sep 2014 04:04:32 +0200
Subject: login: don't overmount /run/user/<UID> on upgrades

Since v215, logind uses a per-user tmpfs for $XDG_RUNTIME_DIR.  When
upgrading from older versions, we don't want to overmount existing user
data, so if that directory already exists, we skip the step where we
mount the per-user tmpfs.

This patch is only necessary for upgrades from pre v215 versions and
should be removed in jessie+1.

Closes: #762041
---
 src/login/logind-user.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/src/login/logind-user.c b/src/login/logind-user.c
index f4c4490..961b579 100644
--- a/src/login/logind-user.c
+++ b/src/login/logind-user.c
@@ -20,6 +20,7 @@
 ***/
 
 #include <sys/mount.h>
+#include <sys/stat.h>
 #include <string.h>
 #include <unistd.h>
 #include <errno.h>
@@ -306,6 +307,7 @@ int user_load(User *u) {
 static int user_mkdir_runtime_path(User *u) {
         char *p;
         int r;
+        struct stat sb;
 
         assert(u);
 
@@ -319,6 +321,12 @@ static int user_mkdir_runtime_path(User *u) {
         } else
                 p = u->runtime_path;
 
+        /* If the directory already exists, don't mount a tmpfs over it */
+        if (stat(p, &sb) == 0 && S_ISDIR(sb.st_mode)) {
+               u->runtime_path = p;
+               return 0;
+        }
+
         if (path_is_mount_point(p, false) <= 0) {
                 _cleanup_free_ char *t = NULL;
 
